tbmaster.1
.\"-
.\" Copyright (c) 2003-2014 Dag-Erling Smørgrav
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd March 24, 2014
.Dt TBMASTER 1
.Os
.Sh NAME
.Nm tbmaster
.Nd manage tinderbox runs
.Sh SYNOPSIS
.Nm
.Op options
.Sh DESCRIPTION
The
.Nm
script manages
.Xr tinderbox 1
runs, generates log summaries, and mails out failure reports.
.Pp
The following options are recognized:
.Bl -tag -width 12n
.It Fl a , Fl -abbreviate
Abbreviate path names in the log file.
Every occurrence of the source directory will be replaced with
.Dq /src ,
and every occurrence of the object directory with
.Dq /obj .
The resulting output should be exactly the same as if the source and
object directories really were
.Pa /src
and
.Pa /obj ,
respectively.
.It Fl c Ar NAME , Fl -config Ns = Ns Ar NAME
The name of the configuration to use.
If specified multiple times, all listed configurations will be run in
sequence.
The default value is the hostname minus the domain part.
.It Fl d , Fl -dump
Dumps the configuration and exits without running the tinderbox.
.It Fl e Ar DIR , Fl -etcdir Ns = Ns Ar DIR
The directory where configuration files are located.
The default value is
.Pa $HOME/etc .
.It Fl h Ar HOSTNAME , Fl -hostname Ns = Ns Ar HOSTNAME
The name of the host running the tinderbox, used in logs and reports.
Can be overridden by the
.Va HOSTNAME
configuration variable.
The default value is that reported by
.Xr uname 1 .
.It Fl l Ar FILE , Fl -lockfile Ns = Ns Ar FILE
The name of a file to lock upon startup.
If the lock is already held by another process,
.Nm
will terminate immediately rather than block.
.It Fl n Ar NUMBER , Fl -ncpu Ns = Ns Ar NUMBER
The maximum number of concurrent builds to run.
Can be overridden by the
.Va NCPU
configuration variable.
The default value is the number of cores in the machine, or 1 if that
number could not be determined.
See the
.Sx Concurrency
section for additional information.
.El
.Ss Configuration files
The
.Nm
script uses named configurations located in individual files named for
the setup they describe, with a
.Pa .rc
suffix.
For instance, the
.Dq snoosnoo
configuration is contained in
.Pa snoosnoo.rc .
.Pp
In addition,
.Nm
attempts to read
.Pa default.rc
and
.Pa site.rc
before and after the actual configuration file, respectively; thus,
they can be used to specify default values shared by multiple
configurations and to override the values set in the individual
configuration files.
.Pp
Each configuration file consists of a list of single- or
multiple-value variable assignments:
.Bl -tag
.It Va variable No = Ar value
Assigns
.Ar value
to the single-value variable
.Va variable .
.It Va variable No = Ar value1 Op No , Ar value2 ...
Assigns
.Ar value1 ,
.Ar value2
etc. to the multi-value variable
.Va variable .
.It Va variable No += Ar value3 Op No , Ar value4 ...
Appends
.Ar value3 ,
.Ar value4
etc. to the multi-value variable
.Va variable .
.It Va variable No -= Ar value5 Op No , Ar value6 ...
Removes
.Ar value5 ,
.Ar value6
etc. from the multi-value variable
.Va variable .
.El
.Pp
Whitespace around the assigment operator and around the commas
separating multiple values is optional.
.Pp
Blank lines are ignored, as is anything following a hash sign
.Pq Sq # .
.Pp
Additionally,
.Cm include
statements can be used to include one configuration in another:
.Bl -tag
.It Cm include Ar otherconfig
.El
.Ss Configuration variables
Below is a list of the configuration variables
.Nm
recognizes and their semantics.
.Pp
Note that many of these variables are passed on as command-line
arguments to
.Xr tinderbox 1 ,
which may provide its own default values for variables which are left
undefined by
.Nm .
.Pp
Some variables are read-only and are provided so that other variables
may include them.
For instance, a common idiom is to derive
.Va OBJDIR
from a combination of
.Va BRANCH ,
.Va ARCH
and
.Va MACHINE .
.Bl -tag -width 12n
.It ARCH
.Pq Vt single, read-only
The architecture currently being built for.
.It BRANCH
.Pq Vt single, read-only
The branch currently being built.
.It BRANCHES
.Pq Vt multiple
A list of source branches to build.
The default value is
.Dq HEAD .
.It CFLAGS
.Pq Vt single
The desired value for the
.Va CFLAGS
.Xr make 1
variable.
This is equivalent to specifying
.Va CFLAGS
in
.Va ENV .
No default value.
.It COPTFLAGS
.Pq Vt single
The desired value for the
.Va COPTFLAGS
.Xr make 1
variable.
This is equivalent to specifying
.Va COPTFLAGS
in
.Va ENV .
No default value.
.It COMMENT
.Pq Vt single
A terse comment describing the setup.
No default value.
.It ENV
.Pq Vt multiple
A list of environment variables to pass to the
.Xr tinderbox 1
script.
Each value is the name and value of an environment variable, separated
by an equal sign
.Pq Sq = .
No default value.
.It HOME
.Pq Vt single, read-only
The current user's home directory, as specified by the
.Ev HOME
environment variable.
Note that it will not be defined unless it passes some simple sanity
checks.
.It HOSTNAME
.Pq Vt single
The name of the host running the tinderbox.
.It JOBS
The maximum number of concurrent
.Xr make 1
processes to run within each build.
No default value.
See the
.Sx Concurrency
section for additional information.
.It LOGDIR
.Pq Vt single
The location of the log directory.
The default value is
.Pa ${SANDBOX}/logs .
.It MACHINE
.Pq Vt single, read-only
The machine currently being built for.
.It NCPU
.Pq Vt single
The maximum number of concurrent builds to run.
No default value.
See the
.Sx Concurrency
section for additional information.
.It OBJDIR
.Pq Vt single
The object directory.
No default value.
.It OPTIONS
.Pq Vt multiple
A list of additional options to pass to the
.Xr tinderbox 1
script.
No default value.
.It PATCH
.Pq Vt single
The file name (either absolute, or relative to the sandbox directory)
of the patch to apply if the
.Dq patch
command is specified.
No default value.
.It PLATFORMS
.Pq Vt multiple
Which architectures and machines to build for.
Each value is the name of an architecture, optionally followed by a
forward slash
.Pq Sq /
and a machine name.
If the machine name is not specified, it is assumed to be identical to
the architecture name.
The default value is
.Dq i386 .
.It RECIPIENT
.Pq Vt multiple
The addresses to which failure reports should be mailed.
The default value is
.Dq ${SENDER} .
.Pp
To avoid unintentional spamming,
.Nm
will strip recipients in the
.Li freebsd.org
domain from this list unless the correct magic sauce is used.
.It SANDBOX
.Pq Vt single
The location of the sandbox directory.
The default value is
.Pa /tmp/tinderbox .
.It SENDER
.Pq Vt single
The envelope sender to use when mailing out failure reports.
This should be a single email address.
No default value.
.It SRCDIR
.Pq Vt single
The source directory.
No default value.
.Pp
Normally, a separate directory within the sandbox will be used for
each build.
Using a shared source directory for all builds reduces disk space
requirements and speeds up the build.
Note that it is generally not a good idea to combine this with any of
the
.Dq clean ,
.Dq cleansrc
or
.Dq update
targets.
.It SUBJECT
.Pq Vt single
The subject to use on failure reports.
The default value is
.Dq Tinderbox failure on ${arch}/${machine} .
.It SVNBASE
.Pq Vt single
The URL to the base of the Subversion repository.
No default value.
.It TARGETS
.Pq Vt multiple
A list of targets (commands) to pass to the
.Xr tinderbox 1
script.
The default value is
.Dq update, world .
.It TIMEOUT
.Pq Vt single
The number of seconds after which each tinderbox invocation will time
out.
No default value.
.It TINDERBOX
.Pq Vt single
The location of the
.Xr tinderbox 1
script.
The default value is
.Dq ${HOME}/bin/tinderbox .
.It URLBASE
.Pq Vt single
If defined, a URL constructed by appending the file name of the full
log file to the expanded value of this variable will be included in
failure reports.
There is no default value.
.El
.Ss Variable Substitution
All configuration variables are subject to variable substitution
immediately before use:
.Bl -bullet
.It
If a single-value variable contains substrings of the form
.Va ${VAR}
or
.Va ${var} ,
those substrings are replaced with the values of the corresponding
variables, after recursive substitution.
The difference between the first and the second form is that the
latter is converted to lower-case before use.
For instance,
.Dq ${BRANCH}
might expand to
.Dq RELENG_4
while
.Dq ${branch}
would expand to
.Dq releng_4 .
.It
If a single-value varaible contains substrings of the form
.Va $ENV{VAR} ,
those substrings are replaced with the values of the corresponding
environment variables.
Use this with care.
.It
If an element of a multiple-value variable is of the form
.Va ${VAR}
or
.Va ${var}
and the corresponding variable is a multiple-value variable, recursive
substitution is first performed on that variable, and the resulting

Changes
This file documents the revision history for Perl extension FBCE.

0.01  2010-06-09 23:31:03
        - initial revision, generated by Catalyst


tbmaster.pl
#!/usr/bin/perl -Tw
#-
# Copyright (c) 2003-2014 Dag-Erling Smørgrav
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
# $FreeBSD$
#

use v5.10.1;
use strict;
use Fcntl qw(:DEFAULT :flock);
use POSIX;
use Getopt::Long;
use Storable qw(dclone);

my $VERSION	= "2.21";
my $COPYRIGHT	= "Copyright (c) 2003-2014 Dag-Erling Smørgrav. " .
		  "All rights reserved.";

my $BACKLOG	= 20;

my $abbreviate;			# Abbreviate path names in log file
my @configs;			# Names of requested configations
my $dump;			# Dump configuration and exit
my $etcdir;			# Configuration directory
my $lockfile;			# Lock file name
my $lock;			# Lock file descriptor
my $hostname;			# Hostname
my $ncpu;			# Number of CPUs
my %platforms;			# Specific platforms to build

my %INITIAL_CONFIG = (
    'BRANCHES'	=> [ 'HEAD' ],
    'CFLAGS'	=> '',
    'COPTFLAGS'	=> '',
    'COMMENT'	=> '',
    'ENV'	=> [ ],
    'HOSTNAME'	=> '',
    'JOBS'	=> '',
    'LOGDIR'	=> '${SANDBOX}/logs',
    'NCPU'	=> '',
    'OBJDIR'	=> '',
    'OPTIONS'	=> [ ],
    'PATCH'	=> '',
    'PLATFORMS'	=> [ 'i386' ],
    'RECIPIENT'	=> [ '${SENDER}' ],
    'SANDBOX'	=> '/tmp/tinderbox',
    'SENDER'	=> '',
    'SRCDIR'	=> '',
    'SUBJECT'	=> 'Tinderbox failure on ${arch}/${machine}',
    'SVNBASE'	=> '',
    'TARGETS'	=> [ 'update', 'world' ],
    'TIMEOUT'   => '',
    'TINDERBOX'	=> '${HOME}/bin/tinderbox',
    'URLBASE'	=> '',
);
my %NUMERIC_OPTIONS = map { $_ => 1 } qw(JOBS NCPU TIMEOUT);
my %PATHNAME_OPTIONS =
    map { $_ => 1 } qw(LOGDIR OBJDIR PATCH SANDBOX SRCDIR TINDERBOX);
my %WORD_OPTIONS = map { $_ => 1 } qw(PLATFORMS TARGETS);
my %CONFIG;

#
# Expand a path
#
sub realpath($;$);
sub realpath($;$) {
    my $path = shift;
    my $base = shift || "";

    my $realpath = ($path =~ m|^/|) ? "" : $base;
    foreach my $part (split('/', $path)) {
	if ($part eq '' || $part eq '.') {
	    # nothing
	} elsif ($part eq '..') {
	    $realpath =~ s|/[^/]+$||
		or die("'$path' is not a valid path relative to '$base'\n");
	} elsif (-l "$realpath/$part") {
	    my $target = readlink("$realpath/$part")
		or die("unable to resolve symlink '$realpath/$part': $!\n");
	    $realpath = realpath($target, $realpath);
	} else {
	    $part =~ m/^([\w.-]+)$/
		or die("unsafe path '$realpath/$part'\n");
	    $realpath .= "/$1";
	}
    }
    return $realpath;
}

#
# Perform variable expansion
#
sub expand($);
sub expand($) {
    my $key = shift;

    return "??$key??"
	unless exists($CONFIG{uc($key)});
    my $value = $CONFIG{uc($key)};
    my @elements = ref($value) ? @{$value} : $value;
    my @expanded;
    while (@elements) {
	my $elem = shift(@elements);
	if (ref($elem)) {
	    # prepend to queue for further processing
	    unshift(@elements, @{$elem});
	} elsif ($elem =~ m/^\%\%(\w+)\%\%$/ || $elem =~ m/^\%\{(\w+)\}$/) {
	    # prepend to queue for further processing
	    # note - can expand to a list
	    unshift(@elements, expand($1));
	} else {
	    $elem =~ s/\$ENV\{(\w+)\}/$ENV{$1}/g;
	    $elem =~ s/\%\%(\w+)\%\%/expand($1)/eg;
	    $elem =~ s/\$\{(\w+)\}/expand($1)/eg;
	    push(@expanded, $elem);
	}
    }

    # Upper / lower case
    if ($key !~ m/[A-Z]/) {
	@expanded = map { lc($_) } @expanded;
    }

    # Validate and untaint expanded value(s)
    if ($NUMERIC_OPTIONS{uc($key)}) {
	@expanded = map {
	    m/^(\d+|)$/
		or die("invalid value for numeric variable $key: $_\n");
	    $1
	} @expanded;
    } elsif ($PATHNAME_OPTIONS{uc($key)}) {
	@expanded = map {
	    m@^((?:/+[\w.-]+)+/*|)$@
		or die("invalid value for pathname variable $key: $_\n");
	    $1
	} @expanded;
    } elsif ($WORD_OPTIONS{uc($key)}) {
	@expanded = map {
	    # hack - support not only "word" but also "word/word" so
	    # platform designations will pass the test.
	    m@^([\w.-]+(?:/[\w.-]+)?|)$@
		or die("invalid value for word variable $key: $_\n");
	    $1
	} @expanded;
    }

    # Verify single / multiple and return result
    if (ref($value)) {
	return @expanded;
    } elsif (@expanded != 1) {
	die("expand($key): expected one value, got ", int(@expanded), "\n");
    } else {
	return $expanded[0];
    }
}

#
# Reset the configuration to initial values
#
sub clearconf() {

    %CONFIG = %{dclone(\%INITIAL_CONFIG)};
}

#
# Read in a configuration file
#
sub readconf($);
sub readconf($) {
    my $fn = shift;

    open(my $fh, '<', $fn)
	or return undef;
    my $line = "";
    my $n = 0;
    while (<$fh>) {
	++$n;
	chomp();
	s/\s*(\#.*)?$//;
	$line .= $_;
	if (length($line) && $line !~ s/\\$/ /) {
	    if ($line =~ m/^include\s+([\w-]+)$/) {
		readconf("$1.rc")
		    or die("$fn: include $1: $!\n");
	    } elsif ($line =~ m/^(\w+)\s*([+-]?=)\s*(.*)$/) {
		my ($key, $op, $val) = (uc($1), $2, $3);
		$val = ''
		    unless defined($val);
		die("$fn: $key is not a known keyword on line $n\n")
		    unless (exists($CONFIG{$key}));
		if (ref($CONFIG{$key})) {
		    my @a = split(/\s*,\s*/, $val);
		    foreach (@a) {
			s/^\'([^\']*)\'$/$1/;
		    }
		    if ($op eq '=') {
			$CONFIG{$key} = \@a;
		    } elsif ($op eq '+=') {
			push(@{$CONFIG{$key}}, @a);
		    } elsif ($op eq '-=') {
			my %a = map { $_ => $_ } @a;
			@{$CONFIG{$key}} =
			    grep { !exists($a{$_}) } @{$CONFIG{$key}};
		    } else {
			die("can't happen\n");
		    }
		} else {
		    $val =~ s/^\'([^\']*)\'$/$1/;
		    if ($op eq '=') {
			$CONFIG{$key} = $val;
		    } elsif ($op eq '+=' || $op eq '-=') {
			die("$fn: $key is not an array on line $n\n");
		    } else {
			die("can't happen\n");
		    }
		}
	    } else {
		die("$fn: syntax error on line $n\n")
	    }
	    $line = "";
	}
    }
    close($fh);
    return 1;
}

#
# Record a tinderbox result in the history file
#
sub history($$$) {
    my $start = shift;
    my $end = shift;
    my $success = shift;

    my $history = expand('HOSTNAME') . "\t";
    $history .= expand('CONFIG') . "\t";
    $history .= strftime("%Y-%m-%d %H:%M:%S\t", localtime($start));
    $history .= strftime("%Y-%m-%d %H:%M:%S\t", localtime($end));
    $history .= expand('ARCH') . "\t";
    $history .= expand('MACHINE') . "\t";
    $history .= expand('BRANCH') . "\t";
    $history .= $success ? "OK\n" : "FAIL\n";

    my $fn = expand('LOGDIR') . "/history";
    if (open(my $fh, '>>', $1)) {
	print($fh $history);
	close($fh);
    } else {
	print(STDERR "failed to record result to history file:\n$history\n");
    }
}

#
# Report a tinderbox failure
#
sub report($$$$) {
    my $sender = shift;
    my $recipient = shift;
    my $subject = shift;
    my $message = shift;

    if (!$message) {
	print(STDERR "[empty report, not sent by email]\n\n]");
	return;
    }
    if (length($message) < 128) {
	print(STDERR "[suspiciously short report, not sent by email]\n\n");
	print(STDERR $message);
	return;
    }

    if (open(my $pipe, '|-', qw(/usr/sbin/sendmail -i -t -f), $sender)) {
	print($pipe "Sender: $sender\n");
	print($pipe "From: $sender\n");
	print($pipe "To: $recipient\n");
	print($pipe "Subject: $subject\n");
	print($pipe "Precedence: bulk\n");
	print($pipe "\n");
	print($pipe "$message\n");
	close($pipe);
    } else {
	print(STDERR "[failed to send report by email]\n\n");
	print(STDERR $message);
    }
}

#
# Run the tinderbox
#
sub tinderbox($$$) {
    my $branch = shift;
    my $arch = shift;
    my $machine = shift;

    my $config = expand('CONFIG');
    my $start = time();

    $0 = "tbmaster [$config]: building $branch for $arch/$machine";

    $CONFIG{'BRANCH'} = $branch;
    $CONFIG{'ARCH'} = $arch;
    $CONFIG{'MACHINE'} = $machine;

    # Open log files: one for the full log and one for the summary
    my $logdir = expand('LOGDIR');
    if (!-d $logdir) {
	die("nonexistent log directory: $logdir\n");
    }
    my $logname = "tinderbox-$config-$branch-$arch-$machine";
    my $logbase = "$logdir/$logname";
    my $full;
    if (!open($full, '>', "$logbase.full.$$")) {
	warn("$logbase.full.$$: $!\n");
	return undef;
    }
    select($full);
    $| = 1;
    select(STDOUT);
    my $brief;
    if (!open($brief, '>', "$logbase.brief.$$")) {
	warn("$logbase.brief.$$: $!\n");
	return undef;
    }
    select($brief);
    $| = 1;
    select(STDOUT);

    # Open a pipe for the tinderbox process
    my ($rpipe, $wpipe);
    if (!pipe($rpipe, $wpipe)) {
	warn("pipe(): $!\n");
	unlink("$logbase.brief.$$");
	close($brief);
	unlink("$logbase.full.$$");
	close($full);
	return undef;
    }

    # Fork and start the tinderbox
    my @args = expand('OPTIONS');
    push(@args, "--hostname=" . expand('HOSTNAME'));
    push(@args, "--sandbox=" . realpath(expand('SANDBOX')));
    push(@args, "--srcdir=" . realpath(expand('SRCDIR')))
	if ($CONFIG{'SRCDIR'});
    push(@args, "--objdir=" . realpath(expand('OBJDIR')))
	if ($CONFIG{'OBJDIR'});
    push(@args, "--arch=$arch");
    push(@args, "--machine=$machine");
    push(@args, "--repository=" . expand('REPOSITORY'))
	if ($CONFIG{'REPOSITORY'});
    push(@args, "--branch=$branch");
    push(@args, "--patch=" . expand('PATCH'))
	if ($CONFIG{'PATCH'});
    push(@args, "--jobs=" . expand('JOBS'))
	if ($CONFIG{'JOBS'});
    push(@args, "--svnbase=" . expand('SVNBASE'))
	if ($CONFIG{'SVNBASE'});
    push(@args, "--timeout=" . expand('TIMEOUT'))
	if ($CONFIG{'TIMEOUT'});
    push(@args, expand('TARGETS'));
    push(@args, expand('ENV'));
    push(@args, "CFLAGS=" . expand('CFLAGS'))
	if ($CONFIG{'CFLAGS'});
    push(@args, "COPTFLAGS=" . expand('COPTFLAGS'))
	if ($CONFIG{'COPTFLAGS'});
    my $pid = fork();
    if (!defined($pid)) {
	warn("fork(): $!\n");
	unlink("$logbase.brief.$$");
	close($brief);
	unlink("$logbase.full.$$");
	close($full);
	return undef;
    } elsif ($pid == 0) {
	close($rpipe);
	open(STDOUT, '>&', $wpipe);
	open(STDERR, '>&', $wpipe);
	$| = 1;
	exec(expand('TINDERBOX'), @args);
	die("exec(): $!\n");
    }

    # Process the output
    close($wpipe);
    my @lines = ();
    my $error = 0;
    my $summary = "";
    my $root = realpath(expand('SANDBOX') . "/$branch/$arch/$machine");
    my $srcdir = realpath(expand('SRCDIR') || "$root/src");
    my $objdir = realpath(expand('OBJDIR') || "$root/obj");
    while (<$rpipe>) {
	if ($abbreviate) {
	    s/\Q$srcdir\E/\/src/go;
	    s/\Q$objdir\E/\/obj/go;
	}
	print($full $_);
	if (/^TB ---/ || /^>>> /) {
	    if ($error) {
		$summary .= join('', @lines);
		print($brief join('', @lines));
		@lines = ();
		$error = 0;
	    }
	    $summary .= $_;
	    print($brief $_);
	    @lines = ();
	    next;
	}
	if (/^\*\*\*( \[.*?\])? (Error code|Stopped|Signal)/ &&
	    !/\(ignored\)/) {
	    $error = 1;
	}
	if (@lines > $BACKLOG && !$error) {
	    shift(@lines);
	    $lines[0] = "[...]\n";
	}
	push(@lines, $_);
    }
    close($rpipe);
    if ($error) {
	$summary .= join('', @lines);
	print($brief join('', @lines));
    }

    # Done...
    if (waitpid($pid, 0) == -1) {
	warn("waitpid(): $!\n");
    } elsif ($? & 0xff) {
	my $msg = "tinderbox caught signal " . ($? & 0x7f) . "\n";
	print($brief $msg);
	print($full $msg);
	$error = 1;
    } elsif ($? >> 8) {
	my $msg = "tinderbox returned exit code " . ($? >> 8) . "\n";
	print($brief $msg);
	print($full $msg);
	$error = 1;
    }
    close($brief);
    close($full);

    my $end = time();

    # Record result in history file
    history($start, $end, !$error);

    # Filter recipients
    my @recipients = expand('RECIPIENT');
    if (!$ENV{'MAGIC_SAUCE'} ||
	$ENV{'MAGIC_SAUCE'} ne 'FREEBSD_TINDERBOX') {
	@recipients = grep { ! m/\@freebsd\.org/i } @recipients;
    }

    # Mail out error reports
    if ($error && @recipients) {
	my $sender = expand('SENDER');
	my $recipient = join(', ', @recipients);
	my $subject = expand('SUBJECT');
	if ($CONFIG{'URLBASE'}) {
	    $summary .= "\n\n" . expand('URLBASE') . "$logname.full";
	}
	report($sender, $recipient, $subject, $summary);
    }

    rename("$logbase.full.$$", "$logbase.full");
    rename("$logbase.brief.$$", "$logbase.brief");
}

#
# Open and lock a file reliably
#
sub open_locked($;$$) {
    my $fn = shift;		# File name
    my $mode = shift;		# Open mode
    my $perm = shift;		# File permissions

    my $fh;			# File handle
    my (@sb1, @sb2);		# File status

    for (;; close($fh)) {
	open($fh, $mode, $fn)
	    or last;
	if (!(@sb1 = stat($fh))) {
	    # Huh? shouldn't happen
	    last;
	}
	if (!flock($fh, LOCK_EX|LOCK_NB)) {
	    # A failure here means the file can't be locked, or
	    # something really weird happened, so just give up.
	    last;
	}
	if (!(@sb2 = stat($fn))) {
	    # File was pulled from under our feet, though it may
	    # reappear in the next pass
	    next;
	}
	if ($sb1[0] != $sb2[0] || $sb1[1] != $sb2[1]) {
	    # File changed under our feet, try again
	    next;
	}
	chmod($fh, $perm)
	    if defined($perm);
	return $fh;
    }
    close($fh);
    return undef;
}

#
# Print a usage message and exit
#
sub usage() {

    (my $self = $0) =~ s|^.*/||;
    print(STDERR "This is the FreeBSD tinderbox manager, version $VERSION.
$COPYRIGHT

Usage:
  $self [options] [parameters]

Options:
  -d, --dump                    Dump the processed configuration

Parameters:
  -a, --abbreviate		Abbreviate path names in log file
  -c, --config=NAME             Configuration name
  -e, --etcdir=DIR              Configuration directory
  -l, --lockfile=FILE           Lock file name
  -n, --ncpu=NUM                Number of CPUs available

Report bugs to <des\@freebsd.org>.
");
    exit(1);
}

#
# Main loop
#
sub tbmaster($) {
    my $config = shift;

    clearconf();
    readconf('default.rc');
    readconf("$config.rc")
	or die("$config.rc: $!\n");
    readconf('site.rc');
    $CONFIG{'CONFIG'} = $config;
    $CONFIG{'ETCDIR'} = $etcdir;

    if ($dump) {
	foreach my $key (sort(keys(%CONFIG))) {
	    printf("%-12s = ", uc($key));
	    if (ref($CONFIG{$key})) {
		print(join(", ", @{$CONFIG{$key}}));
	    } else {
		print($CONFIG{$key});
	    }
	    print("\n");
	}
	return;
    }

    if (!length(expand('TINDERBOX')) || !-x expand('TINDERBOX')) {
	die("Where is the tinderbox script?\n");
    }

    # Check stop file
    my $stopfile = expand('SANDBOX') . "/stop";
    my @jobs;
    foreach my $branch (expand('BRANCHES')) {
	foreach my $platform (expand('PLATFORMS')) {
	    next if (%platforms && !$platforms{$platform});
	    my ($arch, $machine) = split('/', $platform, 2);
	    $machine = $arch
		unless defined($machine);
	    push(@jobs, [ $branch, $arch, $machine ]);
	}
    }

    # Main loop: start as many concurrent jobs as permitted, then keep
    # starting new jobs as soon as existing jobs terminate, until all
    # jobs have terminated and there are none left in the queue.
    $0 = "tbmaster [$config]: supervisor";
    my %children;
    my $done = 0;
    while (@jobs || keys(%children)) {
	# start more children if we can
	while (@jobs && keys(%children) < expand('NCPU')) {
	    my ($branch, $arch, $machine) = @{shift(@jobs)};
	    if (-e $stopfile || -e "$stopfile.$branch" ||
		-e "$stopfile.$arch" || -e "$stopfile.$arch.$machine") {
		warn("stop file found, skipping $branch $arch/$machine\n");
		next;
	    }
	    my $child = fork();
	    if (!defined($child)) {
		die("fork(): $!\n");
	    } elsif ($child == 0) {
		tinderbox($branch, $arch, $machine);
		exit(0);
	    } else {
		$children{$child} = [ $branch, $arch, $machine ];
	    }
	    warn("forked child $child for $branch $arch/$machine\n");
	}
	$0 = "tbmaster [$config]: supervisor (" .
	    keys(%children) . " running, " .
	    @jobs . " pending, " .
	    $done . " completed)";
	# wait for a child to terminate
	if (keys(%children)) {
	    my $child = wait();
	    if ($child > 0) {
		my ($branch, $arch, $machine) = @{$children{$child}};
		warn("child $child for $branch $arch/$machine terminated\n");
		delete($children{$child});
		++$done;
	    }
	}
    }
}

#
# Read the input from a command
#
sub slurp(@) {
    my @cmdline = @_;

    if (open(my $pipe, '-|', @cmdline)) {
	local $/;
	my $input = <$pipe>;
	close($pipe);
	return $input;
    }
    return undef;
}

#
# Main
#
MAIN:{
    # Set defaults
    $ENV{'TZ'} = "UTC";
    $ENV{'PATH'} = "/usr/bin:/usr/sbin:/bin:/sbin";
    if ($ENV{'HOME'} =~ m/^((?:\/[\w\.-]+)+)\/*$/) {
	$INITIAL_CONFIG{'HOME'} = realpath($1);
	$etcdir = "$1/etc";
	$ENV{'PATH'} = "$1/bin:$ENV{'PATH'}"
	    if (-d "$1/bin");
    }

    # Get options
    {Getopt::Long::Configure("auto_abbrev", "bundling");}
    GetOptions(
	"a|abbreviate!"		=> \$abbreviate,
	"c|config=s"		=> \@configs,
	"d|dump"		=> \$dump,
	"e|etcdir=s"		=> \$etcdir,
	"h|hostname=s"		=> \$hostname,
	"l|lockfile=s"		=> \$lockfile,
	"n|ncpu=i"		=> \$ncpu,
	) or usage();

    # Subsequent arguments are platforms to build
    foreach (@ARGV) {
	if (m/^(\w+(?:\/\w+)?)$/) {
	    $platforms{$1} = 1;
	} else {
	    die("invalid platform: $_\n");
	}
    }

    # Get / check hostname
    if (!$hostname) {
	$hostname = slurp(qw(/usr/bin/uname -n));
    }
    if ($hostname &&
	$hostname =~ m/^\s*([a-z][0-9a-z-]+(?:\.[a-z][0-9a-z-]+)*)\s*$/s) {
	$hostname = $1;
    } else {
	$hostname = 'unknown';
    }
    $INITIAL_CONFIG{'HOSTNAME'} = $hostname;

    # Get / check number of CPUs
    if (!$ncpu) {
	$ncpu = slurp(qw(/sbin/sysctl -n hw.ncpu));
    }
    if ($ncpu && $ncpu =~ m/^\s*(\d+)\s*$/s) {
	$ncpu = int($1);
    } else {
	$ncpu = 1;
    }
    $INITIAL_CONFIG{'NCPU'} = $ncpu;

    # Check options
    if (@configs) {
	@configs = split(/,/, join(',', @configs));
    } else {
	$configs[0] = $hostname;
	chomp($configs[0]);
	$configs[0] =~ s/^(\w+)(\..*)?/$1/;
    }
    if (defined($etcdir)) {
	if ($etcdir !~ m/^([\w\/\.-]+)$/) {
	    die("invalid etcdir\n");
	}
	$etcdir = $1;
	chdir($etcdir)
	    or die("$etcdir: $!\n");
    }
    for (my $n = 0; $n < @configs; ++$n) {
	$configs[$n] =~ m/^([\w-]+)$/
	    or die("invalid config: $configs[$n]\n");
	$configs[$n] = $1;
    }

    # Acquire lock
    if (defined($lockfile)) {
	if ($lockfile !~ m/^([\w\/\.-]+)$/) {
	    die("invalid lockfile\n");
	}
	$lockfile = $1;
	$lock = open_locked($lockfile, '>', 0600)
	    or die("unable to acquire lock on $lockfile\n");
	# Lock will be released upon termination.
    }

    # Run all specified or implied configurations
    foreach my $config (@configs) {
	tbmaster($config);
    }
    exit(0);
}
